## 01. Importing Libraries

In [159]:
import pandas as pd
import numpy as np
import os

## 02. Import Data

In [160]:
# path folder
path = r'/Users/lavinia/Documents/04-2020 Instacart Basket Analysis'

In [161]:
# import orders_products_merged_updated
ords_prods_merge = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_merged_updated.pkl'))

In [162]:
# import df_dep
df_dep = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'df_dep_wrangled.csv'), index_col = False)

## - Subset Dataframe

In [163]:
# create a subset for the first one million entries
df = ords_prods_merge[:1000000]

In [164]:
# check the dimension
df.shape

(1000000, 18)

In [165]:
# check the column names
df.head()

order_id  user_id  order_number  order_day_of_week  order_hour_of_day  \
0   2539329        1             1                  2                  8   
1   2398795        1             2                  3                  7   
2    473747        1             3                  3                 12   
3   2254736        1             4                  4                  7   
4    431534        1             5                  4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   

  product_name  aisle_id  department_id  prices _merge    price_range_loc  \
0         Soda        77              7     9.0   both  Mid-range product   
1         Soda        77              7     9.0   both  Mid-range product   
2         Soda        77              7     9.0   both  Mid-range product   
3         Soda        77              7     9.0   both  Mid-range product   
4         Soda        77              7     9.0   both  Mid-range product   

      busiest_day       busiest_days busiest_period_of_day  
0  Regularly busy  Regular busy days        Average orders  
1  Regularly busy       Slowest days        Average orders  
2  Regularly busy       Slowest days           Most orders  
3      Least busy       Slowest days        Average orders  
4      Least busy       Slowest days           Most orders

## 03. Grouping Data with Pandas

## - Subset Dataframe

In [166]:
# group data based on the "product name" column
df.groupby('product_name')

## 04. Aggregating Data with agg()

## - Subset Dataframe

In [167]:
# calculate the mean of the "order_number" column grouped by the "department_id" column
# 1) using agg() function to return the mean values of "order_number"
df.groupby('department_id').agg({'order_number' : ['mean']})

order_number
                      mean
department_id             
4                18.825780
7                17.472355
13               17.993423
14               19.246334
16               19.463012
17               11.294069
19               19.305237
20               17.599636

In [168]:
# calculate the mean of the "order_number" column grouped by the "department_id" column
# 2) using mean() function 
df.groupby('department_id')['order_number'].mean()

department_id
4     18.825780
7     17.472355
13    17.993423
14    19.246334
16    19.463012
17    11.294069
19    19.305237
20    17.599636
Name: order_number, dtype: float64

In [169]:
# Performing Multiple Aggregations
df.groupby('department_id').agg({'order_number' : ['mean', 'min', 'max']})

order_number        
                      mean min max
department_id                     
4                18.825780   1  99
7                17.472355   1  99
13               17.993423   1  99
14               19.246334   1  99
16               19.463012   1  99
17               11.294069   1  98
19               19.305237   1  99
20               17.599636   1  99

## - Entire Dataframe

## Q2: In this Exercise, you learned how to find the aggregated mean of the “order_number” column grouped by “department_id” for a subset of your dataframe. Now, repeat this process for the entire dataframe.

In [170]:
# calculate the average of order number for each department_id 
df_result = ords_prods_merge.groupby('department_id').agg({'order_number': ['mean']}).sort_values(('order_number', 'mean'), ascending = False).reset_index()

In [171]:
# merge the above result using the merge() function to know the department name
df_result_name = df_result.merge(df_dep, on = 'department_id', how = 'left')

/opt/anaconda3/lib/python3.8/site-packages/pandas/core/reshape/merge.py:648: UserWarning: merging between different levels can give an unintended result (2 levels on the left,1 on the right)
  warnings.warn(msg, UserWarning)


In [172]:
print(df_result_name)

    department_id  (department_id, )  (order_number, mean)       department
0              21                 21             22.902379          missing
1              10                 10             20.197148             bulk
2              18                 18             19.310397           babies
3               4                  4             17.811403          produce
4              16                 16             17.665606       dairy eggs
5               2                  2             17.277920            other
6               7                  7             17.225802        beverages
7              19                 19             17.177343           snacks
8               3                  3             17.170395           bakery
9              14                 14             16.773669        breakfast
10             13                 13             16.583536           pantry
11             20                 20             16.473447             deli
12          

## Q3: Analyze the result. How do the results for the entire dataframe differ from those of the subset? Include your comments in a markdown cell below the executed code.

Department with ID 21 (missing) has the highest average order number, followed by ID 10 (bulk) and ID 18 (babies) as the top selling departments. Compared to the subset one, the whole dataframe has a lower average number than the subset one. except for ID 17 (household).

## Q4: Follow the instructions in the exercise for creating a loyalty flag for existing customers usin the transform( ) and loc ( ) functions 

## 05.  Aggregating Data with transform( ) 

In [173]:
# create a new column containing maximum frequecy of the "order_number" column. This function ensures that the resulting series has the same index with the original dataframe
ords_prods_merge['max_order'] = ords_prods_merge.groupby('user_id')['order_number'].transform(np.max)

In [174]:
ords_prods_merge[ords_prods_merge['user_id'] ==1].head(15)

order_id  user_id  order_number  order_day_of_week  order_hour_of_day  \
0       2539329        1             1                  2                  8   
1       2398795        1             2                  3                  7   
2        473747        1             3                  3                 12   
3       2254736        1             4                  4                  7   
4        431534        1             5                  4                 15   
5       3367565        1             6                  2                  7   
6        550135        1             7                  1                  9   
7       3108588        1             8                  1                 14   
8       2295261        1             9                  1                 16   
9       2550362        1            10                  4                  8   
35791   2539329        1             1                  2                  8   
51726   2539329        1             1                  2                  8   
51727   2398795        1             2                  3                  7   
51728    473747        1             3                  3                 12   
51729   2254736        1             4                  4                  7   

       days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                         NaN         196                  1          0   
1                        15.0         196                  1          1   
2                        21.0         196                  1          1   
3                        29.0         196                  1          1   
4                        28.0         196                  1          1   
5                        19.0         196                  1          1   
6                        20.0         196                  1          1   
7                        14.0         196                  2          1   
8                         0.0         196                  4          1   
9                        30.0         196                  1          1   
35791                     NaN       14084                  2          0   
51726                     NaN       12427                  3          0   
51727                    15.0       12427                  3          1   
51728                    21.0       12427                  2          1   
51729                    29.0       12427                  2          1   

                                  product_name  aisle_id  department_id  \
0                                         Soda        77              7   
1                                         Soda        77              7   
2                                         Soda        77              7   
3                                         Soda        77              7   
4                                         Soda        77              7   
5                                         Soda        77              7   
6                                         Soda        77              7   
7                                         Soda        77              7   
8                                         Soda        77              7   
9                                         Soda        77              7   
35791  Organic Unsweetened Vanilla Almond Milk        91             16   
51726                      Original Beef Jerky        23             19   
51727                      Original Beef Jerky        23             19   
51728                      Original Beef Jerky        23             19   
51729                      Original Beef Jerky        23             19   

       prices _merge    price_range_loc     busiest_day       busiest_days  \
0         9.0   both  Mid-range product  Regularly busy  Regular busy days   
1         9.0   both  Mid-range product  Regularly busy       Slowest days   
2         9.0   both  Mid-range product  Regularly busy       Slowest days   
3         9.

In [175]:
# command to tell panda not to limit maximum number of rows to display
pd.options.display.max_rows = None 

## 06 Deriving Column with loc( )

In [176]:
# label customer as 'Loyal customer' when the max order the customer has made is over 40
ords_prods_merge.loc[ords_prods_merge['max_order'] > 40, 'loyalty_flag'] = 'Loyal customer'

In [177]:
# label customer as 'Regular customer' when the max order the customer has made is over 10 but less than or equal to 40
ords_prods_merge.loc[(ords_prods_merge['max_order'] <= 40) & (ords_prods_merge['max_order'] > 10), 'loyalty_flag'] = 'Regular customer'

In [178]:
# label customer as 'New customer' when the max order customer has made is less than or equal to 10
ords_prods_merge.loc[ords_prods_merge['max_order'] <= 10, 'loyalty_flag'] = 'New customer'

In [179]:
# print the frequency of your new "loyalty_flag" column
ords_prods_merge['loyalty_flag'].value_counts(dropna = False)

Regular customer    15876776
Loyal customer      10284093
New customer         6243990
Name: loyalty_flag, dtype: int64

In [180]:
# head function
ords_prods_merge[['user_id', 'loyalty_flag', 'order_number']].head(100)

user_id      loyalty_flag  order_number
0         1      New customer             1
1         1      New customer             2
2         1      New customer             3
3         1      New customer             4
4         1      New customer             5
5         1      New customer             6
6         1      New customer             7
7         1      New customer             8
8         1      New customer             9
9         1      New customer            10
10       15  Regular customer            15
11       15  Regular customer            17
12       15  Regular customer            18
13       15  Regular customer            21
14       15  Regular customer            22
15       19      New customer             2
16       19      New customer             5
17       19      New customer             7
18       21  Regular customer            10
19       31  Regular customer            10
20       31  Regular customer            17
21       43  Regular customer             6
22       43  Regular customer             9
23       52  Regular customer             1
24       52  Regular customer             2
25       52  Regular customer             3
26       52  Regular customer             5
27       52  Regular customer             7
28       52  Regular customer             9
29       52  Regular customer            10
30       52  Regular customer            11
31       52  Regular customer            12
32       52  Regular customer            15
33       52  Regular customer            17
34       52  Regular customer            18
35       52  Regular customer            19
36       52  Regular customer            21
37       67  Regular customer             1
38       67  Regular customer             2
39       67  Regular customer             3
40       67  Regular customer             4
41       67  Regular customer             5
42       67  Regular customer             6
43       67  Regular customer             8
44       67  Regular customer             9
45       67  Regular customer            10
46       67  Regular customer            11
47       67  Regular customer            12
48       67  Regular customer            13
49       67  Regular customer            14
50       67  Regular customer            16
51       67  Regular customer            18
52       67  Regular customer            19
53       67  Regular customer            21
54       67  Regular customer            22
55       67  Regular customer            23
56       81      New customer             3
57       81      New customer             5
58       82  Regular customer             1
59       82  Regular customer             3
60       82  Regular customer             4
61       82  Regular customer             6
62       82  Regular customer             9
63       82  Regular customer            10
64       82  Regular customer            15
65       82  Regular customer            18
66       82  Regular customer            19
67       98  Regular customer             3
68       98  Regular customer             7
69       98  Regular customer             9
70       98  Regular customer            11
71       98  Regular customer            12
72       98  Regular customer            13
73       98  Regular customer            14
74      109      New customer             3
75      120      New customer             1
76      120      New customer             3
77      185      New customer             2
78      195    Loyal customer             1
79      195    Loyal customer             8
80      195    Loyal customer            19
81      195    Loyal customer            22
82      195    Loyal customer            25
83      195    Loyal customer            31
84      195    Loyal customer            34
85      195    Loyal customer            37
86      195    Loyal customer            40
87      195    Loyal customer            47
88      195    Loyal customer            52
89      195    Loyal customer            57


## Q5: The marketing team at Instacart wants to know whether there’s a difference between the spending habits of the three types of customers you identified. Use the loyalty flag you created and check the basic statistics of the product prices for each loyalty category (Loyal Customer, Regular Customer, and New Customer). What you’re trying to determine is whether the prices of products purchased by loyal customers differ from those purchased by regular or new customers

In [181]:
# group ords_prods_merge dataframe by loyalty_flag
ords_prods_merge.groupby('loyalty_flag')['prices'].describe()

count       mean         std  min  25%  50%   75%  \
loyalty_flag                                                               
Loyal customer    10284093.0  10.386336  328.017787  1.0  4.2  7.4  11.2   
New customer       6243990.0  13.294670  597.560299  1.0  4.2  7.4  11.3   
Regular customer  15876776.0  12.495717  539.720919  1.0  4.2  7.4  11.3   

                      max  
loyalty_flag               
Loyal customer    99999.0  
New customer      99999.0  
Regular customer  99999.0

Observation: for each loyalty group of customers, Loyal customers have a lower mean prices of products than New or Regular customers while the regular customers have a slightly lower mean prices than New customers. However, the prices for this group are actually quite spread out due to high standard deviation value, especially for New and Regular customers at 597.5 and 539.7, respectively. Because of the large variability in prices within each group, there is a possibility of significant overlap in prices between the groups, for instance there maybe many loyal customers that spend more than some New or Regular customers despite the lower mean price for Loyal customers.

## Q6: The team now wants to target different types of spenders in their marketing campaigns. This can be achieved by looking at the prices of the items people are buying. Create a spending flag for each user based on the average price across all their orders using the following criteria:
- If the mean of the prices of products purchased by a user is lower than 10, then flag them as a “Low spender.”
- If the mean of the prices of products purchased by a user is higher than or equal to 10, then flag them as a “High spender.”

## Aggregating Data with transform( ) 

In [182]:
# create a new column containing average price across all the user orders
ords_prods_merge['avg_price_order'] = ords_prods_merge.groupby('user_id')['prices'].transform(np.mean)

## Deriving Column with loc( )

In [183]:
# label customer as 'Low spender' when the avg_price_order is lower than 10
ords_prods_merge.loc[ords_prods_merge['avg_price_order'] < 10, 'spending_flag'] = 'Low spender'

In [184]:
# label customer as 'High Spender' when the avg_price_order is higher than or equal to 10
ords_prods_merge.loc[ords_prods_merge['avg_price_order'] >= 10, 'spending_flag'] = 'High spender'

In [185]:
# use head() to display the first five sets of the df
ords_prods_merge.head()

order_id  user_id  order_number  order_day_of_week  order_hour_of_day  \
0   2539329        1             1                  2                  8   
1   2398795        1             2                  3                  7   
2    473747        1             3                  3                 12   
3   2254736        1             4                  4                  7   
4    431534        1             5                  4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   

  product_name  ...  prices  _merge    price_range_loc     busiest_day  \
0         Soda  ...     9.0    both  Mid-range product  Regularly busy   
1         Soda  ...     9.0    both  Mid-range product  Regularly busy   
2         Soda  ...     9.0    both  Mid-range product  Regularly busy   
3         Soda  ...     9.0    both  Mid-range product      Least busy   
4         Soda  ...     9.0    both  Mid-range product      Least busy   

        busiest_days busiest_period_of_day max_order  loyalty_flag  \
0  Regular busy days        Average orders        10  New customer   
1       Slowest days        Average orders        10  New customer   
2       Slowest days           Most orders        10  New customer   
3       Slowest days        Average orders        10  New customer   
4       Slowest days           Most orders        10  New customer   

   avg_price_order spending_flag  
0         6.367797   Low spender  
1         6.367797   Low spender  
2         6.367797   Low spender  
3         6.367797   Low spender  
4         6.367797   Low spender  

[5 rows x 22 columns]

In [186]:
# print the frequency of "marketing_flag" column
ords_prods_merge['spending_flag'].value_counts(dropna = False)

Low spender     31770742
High spender      634117
Name: spending_flag, dtype: int64

## Q7: In order to send relevant notifications to users within the app (for instance, asking users if they want to buy the same item again), the Instacart team wants you to determine frequent versus non-frequent customers. Create an order frequency flag that marks the regularity of a user’s ordering behavior according to the median in the “days_since_prior_order” column. The criteria for the flag should be as follows:
- If the median of “days_since_prior_order” is higher than 20, then the customer should be labeled a “Non-frequent customer.”
- If the median is higher than 10 and lower than or equal to 20, then the customer should be labeled a “Regular customer.”
- If the median is lower than or equal to 10, then the customer should be labeled a “Frequent customer.”


## Aggregating Data with transform( ) 

In [187]:
# create a new column containing the median in the "days_since_prior_order" of the user
ords_prods_merge['median_freq'] = ords_prods_merge.groupby('user_id')['days_since_prior_order'].transform('median')

## Deriving Column with loc( )

In [190]:
# label customer as 'Non-frequent customer' when the median_freq > 20
ords_prods_merge.loc[ords_prods_merge['median_freq'] > 20, 'frequency_flag'] = 'Non-frequent customer'

In [191]:
# label customer as 'Regular customer' when the median_freq > 10 and median_freq <= 20
ords_prods_merge.loc[(ords_prods_merge['median_freq'] > 10) & (ords_prods_merge['median_freq'] <= 20), 'frequency_flag'] = 'Regular customer'

In [192]:
# label customer as 'Frequent customer' when the median_freq <= 10
ords_prods_merge.loc[ords_prods_merge['median_freq'] <= 10, 'frequency_flag'] = 'Frequent customer'

In [194]:
# print the frequenct of frequency_flag
ords_prods_merge['frequency_flag'].value_counts(dropna = False)

Frequent customer        21559853
Regular customer          7208564
Non-frequent customer     3636437
NaN                             5
Name: frequency_flag, dtype: int64

## 07. Exporting Data

In [ ]:
ords_prods_merge.to_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_merged_updated_v2.pkl'))